docs/book/src/tasks/experimental-features/cluster-class/write-clusterclass.md

```diff
@@ -1,3 +1,618 @@
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  variables:
+  - name: imageRepository
+    required: true
+    schema:
+      openAPIV3Schema:
+        type: string
+        description: ImageRepository is the container registry to pull images from.
+        default: k8s.gcr.io
+        example: k8s.gcr.io
+```
+
+<aside class="note">
+
+The following basic types are supported: `string`, `integer`, `number` and `boolean`. We are also 
+supporting complex types, please see the [complex variables](#complex-variables) section.
+
+</aside>
+
+**Defining patches in the ClusterClass**
+
+The variable can then be used in a patch to set a field on a template referenced in the ClusterClass.
+The `selector` specifies on which template the patch should be applied. `jsonPatches` specifies which JSON 
+patches should be applied to that template. In this case we set the `imageRepository` field of the 
+`KubeadmControlPlaneTemplate` to the value of the variable `imageRepository`. For more information 
+please see the [godoc](https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api/cluster.x-k8s.io/ClusterClass/v1beta1#spec-patches-definitions).
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: ClusterClass
+metadata:
+  name: docker-clusterclass-v0.1.0
+spec:
+  ...
+  patches:
+  - name: imageRepository
+    definitions:
+    - selector:
+        apiVersion: controlplane.cluster.x-k8s.io/v1beta1
+        kind: KubeadmControlPlaneTemplate
+        matchResources:
+          controlPlane: true
+      jsonPatches:
+      - op: add
+        path: /spec/template/spec/kubeadmConfigSpec/clusterConfiguration/imageRepository
+        valueFrom:
+          variable: imageRepository
+```
+
+<aside class="note">
+
+* Only fields below `/spec` can be patched.
+* Only `add`, `remove` and `replace` operations are supported.
+* It's only possible to append and prepend to arrays. Insertions at a specific index are 
+  not supported.
+* Be careful, appending or prepending an array variable to an array leads to a nested array
+  (for more details please see this [issue](https://github.com/kubernetes-sigs/cluster-api/issues/5944)).
+
+</aside>
+
+**Setting variable values in the Cluster**
+
+After creating a ClusterClass with a variable definition, the user can now provide a value for 
+the variable in the Cluster as in the example below.
+
+```yaml
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: Cluster
+metadata:
+  name: my-docker-cluster
+spec:
+  topology:
+    ...
+    variables:
+    - name: imageRepository
+      value: k8s.gcr.io
+```
+
+<aside class="note">
+
+If the user does not set the value, the variable is automatically added with its 
+default value by the Cluster webhook, as specified in the corresponding variable 
+definition in the ClusterClass.
+
+</aside>
+
+## Advanced features of ClusterClass with patches
```

Comment:
Proposing a little order reshuffle of the use cases

- MachineDeployment variable overrides
- Builtin variables
- Complex variable types
- Using variables values in JSON patches
- Optional patches

Reply:
Fine for me